File: modules/data/tblrec.lua
sptbl["tblrec"] = {

    files = {
        module = "tblrec.c",
        header = "tblrec.h",
        example = "ex_tblrec.c",
    },

    func = {
        create = "ut_tblrec_create",
        destroy = "ut_tblrec_destroy",
        init = "ut_tblrec_init",
        compute = "ut_tblrec_compute",
    },

    params = {
        mandatory = {
            {
                name = "bar",
                type = "ut_ftbl *",
                description = "",
                default = "N/A"
            },
        },
    },

    modtype = "module",

    description = [[Record a signal to an ftable.
This module will write audio-rate signals to a preallocated utone ftable. 
Every time record is enabled, it will got back to index 0 overwrite any
previous information that was on it. 
]],

    ninputs = 2,
    noutputs = 1,

    inputs = {
        {
            name = "in",
            description = "Input signal."
        },
        {
            name = "trig",
            description = "When non-zero, will toggle recording on/off. Recording will set the playhead back to zero"
        },
    },

    outputs = {
        {
            name = "out",
            description = "Copy of the input signal."
        }
    }

}

File: modules/tblrec.c
#include <stdlib.h>
#include <string.h>
#include "utone.h"

int ut_tblrec_create(ut_tblrec **p)
{
    *p = malloc(sizeof(ut_tblrec));
    return UT_OK;
}

int ut_tblrec_destroy(ut_tblrec **p)
{
    free(*p);
    return UT_OK;
}

int ut_tblrec_init(ut_data *ut, ut_tblrec *p, ut_ftbl *ft)
{
    p->index = 0;
    p->record = 0;
    p->ft = ft;
    return UT_OK;
}

int ut_tblrec_compute(ut_data *ut, ut_tblrec *p, UTFLOAT *in, UTFLOAT *trig, UTFLOAT *out)
{
    if(*trig != 0) {
        if(p->record == 1) {
            p->record = 0;
        } else {
            p->record = 1;
            p->index = 0;
            memset(p->ft->tbl, 0, sizeof(UTFLOAT) * p->ft->size);
        }
    }

    if(p->record) {
        p->ft->tbl[p->index] = *in;
        p->index = (p->index + 1) % p->ft->size;
    }

    *out = *in;
    return UT_OK;
}
